File: src/CSharp/901-1000/914.HasGroupsSizeX.cs

using System;
using System.Linq;
using System.Collections.Generic;

namespace PeefyLeetCode.HasGroupsSizeX
{
public class Solution {
    int gcd(int a, int b)
	{
		if (a % b == 0)
			return b;
		return gcd(b, a % b);
	}

    public bool HasGroupsSizeX(int[] deck) {
        Dictionary<int, int> dic = new Dictionary<int, int>();
        foreach (int d in deck){
            if (dic.ContainsKey(d) == false)
                dic.Add(d, 1);
            else
                dic[d]++;
        }
        var arr = dic.Values.ToArray();
        var min = arr[0];
        if (min < 2)
			return false;
        for (int i = 1; i < arr.Length; ++i)
		{
			min = gcd(min, arr[i]);
			if (min < 2)
				return false;
		}
		return true;
    }
}
}

File: src/Python/901-1000/914.HasGroupsSizeX.py

from collections import Counter

class Solution:
    def hasGroupsSizeX(self, deck):
        """
        :type deck: List[int]
        :rtype: bool
        """
        if len(deck) < 2:
            return False
        counter = Counter(deck)
        for x in range(2, min(counter.values()) + 1):
            if all(map(lambda v : v % x == 0, counter.values())):
                return True
        return False

if __name__ == '__main__':
    solution = Solution()
    print(solution.hasGroupsSizeX([1,2,3,4,4,3,2,1]))
    print(solution.hasGroupsSizeX([1,1,1,2,2,2,3,3]))
    print(solution.hasGroupsSizeX([1]))
    print(solution.hasGroupsSizeX([1,1]))
    print(solution.hasGroupsSizeX([0,0,0,1,1,1,2,2,2]))
else:
    pass


